ipamd/datastore/data_store.go

```diff
@@ -321,9 +323,31 @@ func (ds *DataStore) assignPodIPv4AddressUnsafe(k8sPod *k8sapi.K8SPodInfo) (stri
 }
 
 // GetStats returns statistics
-// it returns total number of IP addresses, number of assigned IP addresses
-func (ds *DataStore) GetStats() (int, int) {
-	return ds.total, ds.assigned
+// it returns total number of IP addresses, number of assigned IP addresses,
+// and number of IPs in "cooling mode"
+func (ds *DataStore) GetStats() (int, int, int) {
+	return ds.total, ds.assigned, ds.getCoolingIPCount()
+}
+
+func (ds *DataStore) getCoolingIPCount() int {
+	count := 0
+	curTime := time.Now()
+	for _, eni := range ds.eniIPPools {
+		for _, addr := range eni.IPv4Addresses {
+			if !addr.Assigned &&
+				!addr.unAssignedTime.IsZero() &&
+				curTime.Sub(addr.unAssignedTime) <= addressCoolingPeriod {
+				count++
+			}
+		}
+	}
+	return count
+}
+
+func (ds *DataStore) withStats(logf func(string, ...interface{}), format string, a ...interface{}) {
+	total, assigned, cooling := ds.GetStats()
```

Comment:
This will have `big performance` impact!  It can NOT loop through the all IP addresses  in ENI for every stats.
Also, the cooling IP only last 30 seconds.  Not sure if this is useful.

Reply:
> This will have `big performance` impact! It can NOT loop through the all IP addresses in ENI for every stats.

I understand your concern and thought about this as well, but in testing this on a cluster using m5.xlarge instances the impact on performance has been negligible. See the following graph. The update is clearly visible in the middle.

![screen shot 2018-10-23 at 10 05 01](https://user-images.githubusercontent.com/2261897/47341986-2ac36400-d6ab-11e8-9ca8-bb5a320e7b3b.png)

> Also, the cooling IP only last 30 seconds. Not sure if this is useful.

I'm making this change because it is useful for us. I spent half a day chasing ghosts, because the logs say (as explained in the PR description) that there's IPs available, but IP allocation fails. Also, you can see on the graph below, that in a cluster that uses CronJobs, some IPs are **always** in cooling mode (the thin purple area at the bottom). 

![screen shot 2018-10-23 at 10 01 35](https://user-images.githubusercontent.com/2261897/47341857-c86a6380-d6aa-11e8-8c31-c3156bddf595.png)
  
However, if you will not accept the change as is, then there's a few alternatives:

1. Increment a cooling stat every time an IP is unallocated and then decrement it 30 seconds later.
  a) This will work now, but is not very future proof. If the behavior for cooling IPs gets more complex, then this has to also be kept up to date, which might be difficult.
2. Re-calculate the stat every time an IP is unallocated and then 30 seconds later, by looping through all IPs.
  a) This will avoid looping through all IPs for every log message, but will still loop twice every  time an IP is unallocated. A compromise between the other two options.

I've gone ahead and implemented alternative nr 1, assuming that that will be your preference.

Reply:
@deiwin , 

 I am worrying about possible too many 30sec timers triggered by each IP-unallocation.  I am also worrying about the  "locking" when trying to decrement in those timeout routines and "locks" can introduce delay for allocating a IP for a pod